Readme.md
# Introduction

Animation Generator is an utility program to generate animated GIF files from Multiburst Sony files provided in JPG format.

# Overview

Many compact cameras from Sony, allow the user to take a set of snapshots, which get then stored into a single JPG file.

As most photo applications don't reckognise this format, the user is not able to make use of them without making use of the camera delivered software.

This is of course a burden for users not making use of platforms supported by the said applications.

# Using

To use this application, you need to drop existing Multiburst images into the application, and select the _Process Images_ button.

# Building

This application was developed in Java. A Netbeans project definition file is also available.

Changes.txt
======== V 0.0.1 Release was released ===========================

2007       (first release) : Made the files available on my own web site. (Paulo Pinto)
2012-11-24 (second release): Pushed the files into GitHub. (Paulo Pinto)


src/org/progtools/DropHandler.java
/*
 * DropHandler.java
 * Processes the droping of files into a Swing component.
 * Copyright (C) 2007  Paulo Pinto
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */
package org.progtools;

import java.awt.datatransfer.DataFlavor;
import java.awt.datatransfer.Transferable;
import java.io.File;
import java.util.Iterator;
import java.util.List;

import javax.swing.DefaultListModel;
import javax.swing.JComponent;
import javax.swing.JList;
import javax.swing.TransferHandler;

/**
 * The class manages the processing of drop messages
 * into a Swing component.
 * Only allows the drop of file objects that have a
 * .jpg/.jpeg extension.
 *
 * @author ppinto
 */
public class DropHandler extends TransferHandler {
    /**
     * Validates if the data that is being dropped is valid.
     * @param comp Where the drop is performed.
     * @param transferFlavors Type of objects being dropped.
     * @return True if the drop is acceptable.
     */
    public boolean canImport(JComponent comp, DataFlavor[] transferFlavors) {
        for (int i = 0; i < transferFlavors.length; i++) {
            if (!transferFlavors[i].isFlavorJavaFileListType()) {
                return false;
            }
        }
        
        // If we reach here all drops are acceptable
        return true;
    }
    
    
    /**
     * Process the objects being dropped.
     * @param comp The component that will receive the data. Must use DefaultListModel as a model.
     * @param t Data being transfered.
     * @returns true if the data was sucessfuly imported.
     */
    @Override
    public boolean importData(JComponent comp, Transferable t) {
        try {
            JList list = (JList) comp;
            DefaultListModel model = (DefaultListModel) list.getModel();
            
            List files= (List) t.getTransferData(DataFlavor.javaFileListFlavor);
            Iterator iter = files.iterator();
            while (iter.hasNext()) {
                File fd = (File) iter.next();
                String pathname = fd.getPath();
                if (isValidFileType(pathname)) {
                    model.addElement(pathname);
                }
            }
            
        } catch (Exception e) {
            return false;
        }
        
        return true;
    }
    
    /**
     * Tells if this is a valid type of file to process.
     * @param filename The complete path to the filename.
     * @returns true if this is a jpeg based image.
     */
    private boolean isValidFileType(String filename) {
        String lowercase = filename.toLowerCase();
        return lowercase.endsWith(".jpg") || lowercase.endsWith(".jpeg");
    }
}


src/org/progtools/AnimGIFGenerator.java
/*
 * AnimGIFGenerator.java
 * An image processor to generate frammed images with copyright messages
 * Copyright (C) 2007  Paulo Pinto
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */
package org.progtools;

import java.awt.Graphics2D;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.util.Iterator;
import javax.imageio.IIOImage;
import javax.imageio.ImageIO;
import javax.imageio.ImageWriter;
import javax.imageio.stream.FileImageOutputStream;

/**
 * A GIF generator from Sony's Multi-Burst image files.
 * @author Paulo
 */
public class AnimGIFGenerator {
    private int imagesPerRow;
    private int imagesPerCol;
    
    
    /**
     * Prepares the class instance to process images of a given size.
     *
     * @param colCount The number of frames per column on the image.
     * @param rowCount The number of frames per row on the image.
     */
    public AnimGIFGenerator(int colCount, int rowCount) {
        imagesPerCol = colCount;
        imagesPerRow = rowCount;
    }
    
    /**
     * Given a pathname to a jpg image, processes it and generates
     * an animated gif in the same directory as the given image.
     * @param pathnameToImage The pathname to the jpg image file.
     * @throws IOException If there is an error reading or writing the image data.
     */
    public void generateGIF(String pathnameToImage) throws IOException{
        // Prepare the amout of frames to generate
        BufferedImage[] frames = getMovieFrames(new File(pathnameToImage));
        
        // Now write the image
        File movieFile = convertFilename(pathnameToImage);
        saveAnimation(movieFile, frames);
    }

    /**
     * Reads the jpg image file and splits the frames into individual image
     * objects.
     *
     * @param filename The pathname to the jpg image file.
     * @returns An array containing the animation frames.
     * @throws IOException If there is an error reading or writing the image data.
     */
    private BufferedImage[] getMovieFrames(File filename) throws IOException {
        // Load the images
        BufferedImage originalImg = ImageIO.read(filename);
        
        // Get the width/height per frame
        int frameWidth = originalImg.getWidth() / imagesPerRow;
        int frameHeight = originalImg.getHeight() / imagesPerCol;
        int frameCount = imagesPerCol * imagesPerRow;
        
        // Prepare the amout of frames to generate
        BufferedImage[] frames = new BufferedImage[frameCount];
        for (int i = 0; i < frames.length; i++) {
            BufferedImage destImg = new BufferedImage(frameWidth, frameHeight, originalImg.getType());
            
            // Now copy the frame into the image
            Graphics2D g = (Graphics2D) destImg.getGraphics();
            
            int x = frameWidth * (i % imagesPerRow);
            int y = frameHeight * (i / imagesPerCol);
            g.drawImage(originalImg, 0, 0, frameWidth, frameHeight, x, y, x + frameWidth, y + frameHeight,null);
            frames[i] = destImg;
        }
        
        return frames;
    }
    
    /**
     * Converts the filename into a file object targeting
     * a file with the same path as the given argument but
     * with a .gif extension.
     *
     * @param filename The filename to convert. Must have an extension
     * @return An instance of File pointing to the new name.
     */
    private File convertFilename(String filename) {
        String newName = null;
        int pos = filename.lastIndexOf('.');
        if (pos > 0) {
            newName = filename.substring(0, pos) + ".gif";
        }
        return new File(newName);
    }
    
    /**
     * Saves a sequence of images as an animated gif.
     *
     * @param filename The name of the file where to write to. It is advisable to end in ".gif".
     * @param frames The images to write to the file as animation frames.
     * @throws IOException If there is an error writing to the destination file.
     */
    private void saveAnimation(File filename, BufferedImage[] frames) throws IOException {
        Iterator<ImageWriter> writers = ImageIO.getImageWritersByFormatName("gif");
        ImageWriter writer = writers.next(); // Always assume GIF is available

        // prepare the sequence writer
        writer.setOutput(new FileImageOutputStream(filename));
        writer.prepareWriteSequence(null);
        
        // write the sequence
        for (int i = 0; i < frames.length; i++) {
            IIOImage img = new IIOImage(frames[i], null, null);
            writer.writeToSequence(img, null);
            
        }
        
        // terminate the sequence writer
        writer.endWriteSequence();
    }
}


src/org/progtools/BackgroundProcessor.java
/*
 * BackgroundProcessor.java
 * Application window for the generator class.
 * Copyright (C) 2007  Paulo Pinto
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

package org.progtools;

import java.io.IOException;
import java.util.Enumeration;
import javax.swing.JFrame;
import javax.swing.JOptionPane;
import javax.swing.JProgressBar;
import javax.swing.SwingUtilities;

/**
 * Swing worker thread to process the image generation in the background.
 * @author Paulo Pinto
 */
public class BackgroundProcessor implements Runnable  {
    private JProgressBar progressBar;
    private Enumeration<?> images;
    private AnimGIFGenerator generator;
    private JFrame owner;
            
    /**
     * Creates a new instance of BackgroundProcessor.
     * @param owner The window that owns the background task.
     * @param progressBar The progress bar to update asynchronously.
     * @param images The set of images to process.
     * @param generator The generator that converts the images to the desired format.
     */
    public BackgroundProcessor(JFrame owner, JProgressBar progressBar, Enumeration<?> images, AnimGIFGenerator generator) {
        this.owner = owner;
        this.progressBar = progressBar;
        this.images = images;
        this.generator = generator;
    }

    /**
     * Implements the thread action
     */
    @Override
    public void run() {
        int count = 1;
        while (images.hasMoreElements()) {
            try {
                updateProgressBar(count++);
                generator.generateGIF((String)images.nextElement());
            } catch (IOException e) {
                JOptionPane.showMessageDialog(owner, e, "Error while processing file", JOptionPane.ERROR_MESSAGE);
            }
        }
        resetState();
    }
    
    /**
     * Updates the user interface progress bar in a thread safe way
     * @param value The current value to be setted on the progress bar.
     */
    private void updateProgressBar(final int value) {
        try {
            SwingUtilities.invokeAndWait(new Runnable() {
                public void run() {
                    progressBar.setValue(value);
                }
            });
        } catch (Exception ex) {
            // Just ignore it for the time being
        }
    }

    /**
     * Sets the cursor and progress bar again to a normal state
     */
    private void resetState() {
        try {
            SwingUtilities.invokeAndWait(new Runnable() {
                public void run() {
                    progressBar.setValue(0);
                    owner.setCursor(null);
                }
            });
        } catch (Exception ex) {
            // Just ignore it for the time being
        }
    }
    
}


src/org/progtools/FileSelector.java
/*
 * FileSelector.java
 * Application window for the generator class.
 * Copyright (C) 2007  Paulo Pinto
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */
package org.progtools;

import java.awt.Cursor;
import java.beans.PropertyChangeEvent;
import java.beans.PropertyChangeListener;
import java.io.IOException;
import java.util.Enumeration;
import javax.swing.DefaultListModel;
import javax.swing.JOptionPane;
import javax.swing.UIManager;

/**
 * The application window.
 * @author  Paulo
 */
public class FileSelector extends javax.swing.JFrame implements PropertyChangeListener  {
    private AnimGIFGenerator generator;
    private DefaultListModel filesToProcess;
    
    /** Creates new form FileSelector */
    public FileSelector() {
        generator = new AnimGIFGenerator(4, 4);
        filesToProcess = new DefaultListModel();
        initComponents();
    }
    
    /** This method is called from within the constructor to
     * initialize the form.
     * WARNING: Do NOT modify this code. The content of this method is
     * always regenerated by the Form Editor.
     */
    // <editor-fold defaultstate="collapsed" desc=" Generated Code ">//GEN-BEGIN:initComponents
    private void initComponents() {
        jProgressBar1 = new javax.swing.JProgressBar();
        jScrollPane1 = new javax.swing.JScrollPane();
        filesList = new javax.swing.JList();
        filesList.setModel(filesToProcess);
        filesList.setTransferHandler(new DropHandler());
        processImgBtn = new javax.swing.JButton();
        exitBtn = new javax.swing.JButton();
        removeSelectionBtn = new javax.swing.JButton();
        progressBar = new javax.swing.JProgressBar();

        setDefaultCloseOperation(javax.swing.WindowConstants.EXIT_ON_CLOSE);
        setTitle("Animation Generator");
        jScrollPane1.setViewportView(filesList);

        processImgBtn.setText("Process Images");
        processImgBtn.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                processImgBtnActionPerformed(evt);
            }
        });

        exitBtn.setText("Exit");
        exitBtn.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                exitBtnActionPerformed(evt);
            }
        });

        removeSelectionBtn.setText("Remove Selection");
        removeSelectionBtn.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                removeSelectionBtnActionPerformed(evt);
            }
        });

        javax.swing.GroupLayout layout = new javax.swing.GroupLayout(getContentPane());
        getContentPane().setLayout(layout);
        layout.setHorizontalGroup(
            layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(layout.createSequentialGroup()
                .addContainerGap()
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.TRAILING)
                    .addComponent(progressBar, javax.swing.GroupLayout.Alignment.LEADING, javax.swing.GroupLayout.DEFAULT_SIZE, 297, Short.MAX_VALUE)
                    .addComponent(jScrollPane1, javax.swing.GroupLayout.Alignment.LEADING, javax.swing.GroupLayout.DEFAULT_SIZE, 297, Short.MAX_VALUE))
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.TRAILING)
                    .addComponent(exitBtn, javax.swing.GroupLayout.Alignment.LEADING, javax.swing.GroupLayout.DEFAULT_SIZE, 117, Short.MAX_VALUE)
                    .addComponent(removeSelectionBtn, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                    .addComponent(processImgBtn, javax.swing.GroupLayout.DEFAULT_SIZE, 117, Short.MAX_VALUE))
                .addContainerGap())
        );
        layout.setVerticalGroup(
            layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(layout.createSequentialGroup()
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                    .addGroup(layout.createSequentialGroup()
                        .addGap(16, 16, 16)
                        .addComponent(processImgBtn)
                        .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                        .addComponent(removeSelectionBtn)
                        .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                        .addComponent(exitBtn))
                    .addGroup(layout.createSequentialGroup()
                        .addContainerGap()
                        .addComponent(jScrollPane1)))
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                .addComponent(progressBar, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE)
                .addContainerGap())
        );
        pack();
    }// </editor-fold>//GEN-END:initComponents
    
    /**
     * Removes all the files the user has selected.
     * @param evt Information about the triggered event.
     */
    private void removeSelectionBtnActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_removeSelectionBtnActionPerformed
        for (int index: filesList.getSelectedIndices()) {
            filesToProcess.remove(index);
        }
    }//GEN-LAST:event_removeSelectionBtnActionPerformed
    
    
    /**
     * Processes the list of available files, generated the animated GIF for each one
     * of them. Assuming they are on the proper format.
     * @param evt Information about the triggered event.
     */
    private void processImgBtnActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_processImgBtnActionPerformed
        // Prepare the interface to give feedback to the user about the
        // progress of the image conversion.
        setCursor(Cursor.getPredefinedCursor(Cursor.WAIT_CURSOR));
        progressBar.setMinimum(0);
        progressBar.setMaximum(filesToProcess.size());
        progressBar.setValue(0);
        Enumeration<?> iter = filesToProcess.elements();
        /*while (iter.hasMoreElements()) {
            try {
                generator.generateGIF((String)iter.nextElement());
            } catch (IOException e) {
                JOptionPane.showMessageDialog(this, e, "Error while processing file", JOptionPane.ERROR_MESSAGE);
            }
        }
        setCursor(null);*/
        BackgroundProcessor task = new BackgroundProcessor(this, progressBar, iter, generator);
        new Thread(task).start();
    }//GEN-LAST:event_processImgBtnActionPerformed
    
    /**
     * Processes the exit button clicked event. Just leaves the applicaton.
     * @param evt Information about the triggered event.
     */
    private void exitBtnActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_exitBtnActionPerformed
        dispose();
    }//GEN-LAST:event_exitBtnActionPerformed
    
    /**
     * Application entry point.
     * @param args the command line arguments
     */
    public static void main(String args[]) {
        try {
            UIManager.setLookAndFeel(
                    UIManager.getSystemLookAndFeelClassName());
            
            java.awt.EventQueue.invokeLater(new Runnable() {
                public void run() {
                    new FileSelector().setVisible(true);
                }
            });
        } catch (Exception e) {
            // Some nasty problem happened, just leave.
            e.printStackTrace();
            System.exit(-1);
        }
    }
    
    /**
     * Called by the background task to notify about the task's progress
     *
     * @param evt The event details.
     */
    public void propertyChange(PropertyChangeEvent evt) {
        if ("progress" == evt.getPropertyName()) {
            int progress = (Integer) evt.getNewValue();
            progressBar.setValue(progress);
        }
    }
    
    // Variables declaration - do not modify//GEN-BEGIN:variables
    private javax.swing.JButton exitBtn;
    private javax.swing.JList filesList;
    private javax.swing.JProgressBar jProgressBar1;
    private javax.swing.JScrollPane jScrollPane1;
    private javax.swing.JButton processImgBtn;
    private javax.swing.JProgressBar progressBar;
    private javax.swing.JButton removeSelectionBtn;
    // End of variables declaration//GEN-END:variables
    
}
